tower/src/ready_cache/cache.rs

```diff
@@ -410,8 +397,13 @@ where
     type Output = Result<(K, S, CancelRx), PendingError<K, S::Error>>;
 
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        let mut fut = self.cancel.as_mut().expect("polled after complete");
-        if let Poll::Ready(r) = Pin::new(&mut fut).poll(cx) {
+        // This MUST return ready as soon as the sender has been notified so
+        // that we don't return a service that has been canceled, so we disable
+        // cooperative scheduling on the receiver. Otherwise, the receiver can
+        // sporadically return pending even though the sender has fired.
+        let mut cancel =
+            tokio::task::unconstrained(self.cancel.as_mut().expect("polled after complete"));
```

Comment:
style nit, take it or leave it: might consider something like this, so that it's very obvious that the comment is referring to the use of `unconstrained`:
```suggestion
        let cancel = self.cancel.as_mut().expect("polled after complete");
        // This MUST return ready as soon as the sender has been notified so
        // that we don't return a service that has been canceled, so we disable
        // cooperative scheduling on the receiver. Otherwise, the receiver can
        // sporadically return pending even though the sender has fired.
        let mut cancel = tokio::task::unconstrained(cancel);
```